### PR TITLE
Introduce a ICON test case

### LINKED ISSUE #631: Identify ICON Stencil Suitable for Translation
Having Introduced weighted reductions and sparse dimensions it is now be possible to translate small sections of the ICON Code. For this task, identify a suitable stencil containing not more than a few reductions, preferably using both sparse dimensions and weights.

DoD: Identify the stencil and write it in dawn using the python SIR, e.g. as a (disabled by default) DawnC compiler test

### PR DESCRIPTION
**Pull request by mroethlin**
## Technical Description

This test introduces a ICON stencil as a prototype test. A detailed write up is found [here](https://docs.google.com/document/d/1-9DF7fqiZqXsM84DBGflLqj-iGGKwUKuT1mkdgsQ0YM/edit?usp=sharing). 

### Resolves / Enhances

Resolves #631  and #633 with some differences:
- The code is generated from IIR, not SIR, for easier of testing
- Both hand rolled and generated code use `mylib`. It was envisioned that the generated code uses Atlas. A prototype for Atlas was implemented as well, but does not produce the same result as the `mylib` versions yet

### Testing

No automated testing. Running both generated and manually written versions of the stencil in the prototype folder with `dbg_out = true` writes results and intermediary results on disk, which can be visualized and compared with suitable tools (e.g. Octave)

### Dependencies

This PR is independent

### CODE REVIEW

File: dawn/prototype/mylibIconLaplace.cpp [hunk 1]
```diff
@@ -0,0 +1,597 @@
...
+  // -1 otherwise
+  for(const auto& v : mesh.vertices()) {
+    int m_sparse = 0;
+    for(const auto& e : v.edges()) {
+      mylib::Vertex testVec =
+          mylib::Vertex(v.vertex(m_sparse).x() - v.x(), v.vertex(m_sparse).y() - v.y(), -1);
+      mylib::Vertex dual = mylib::Vertex(dual_normal_x(*e, level), dual_normal_y(*e, level), -1);
+      edge_orientation_vertex(v, m_sparse, level) =
+          sgn(dot(mylib::Vertex(v.vertex(m_sparse).x() - v.x(), v.vertex(m_sparse).y() - v.y(), -1),
+                  mylib::Vertex(dual_normal_x(*e, level), dual_normal_y(*e, level), -1)));
```
Comment (Stagno) [inline dawn/prototype/mylibIconLaplace.cpp]: can use `testVec` and `dual`

File: dawn/prototype/mylibIconLaplace.cpp [hunk 2]
```diff
@@ -0,0 +1,597 @@
...
+    int m_sparse = 0;
+    for(const auto& e : v.edges()) {
+      geofac_rot(v, m_sparse, level) = dual_edge_length(*e, level) *
+                                       edge_orientation_vertex(v, m_sparse, level) /
+                                       dual_cell_area(v, level);
+
+      // ptr_int%geofac_rot(jv,je,jb) =                &
+      //    & ptr_patch%edges%dual_edge_length(ile,ibe) * &
+      //    & ptr_patch%verts%edge_orientation(jv,jb,je)/ &
+      //    & ptr_patch%verts%dual_area(jv,jb) * REAL(ifac,wp)
```
Comment (Stagno) [inline dawn/prototype/mylibIconLaplace.cpp]: what is `REAL(ifac,wp)`? should we care?

File: dawn/prototype/mylibIconLaplaceDriver.cpp [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+
+  //===------------------------------------------------------------------------------------------===//
+  // computation starts here
+  //===------------------------------------------------------------------------------------------===//
+
+  dawn_generated::cxxnaiveico::icon<mylibInterface::mylibTag>(
+      mesh, k_size, vec, div_vec, rot_vec, nabla2t1_vec, nabla2t2_vec, nabla2_vec,
+      primal_edge_length, dual_edge_length, tangent_orientation, geofac_rot, geofac_div)
+      .run();
+
```
Comment (Stagno) [inline dawn/prototype/mylibIconLaplaceDriver.cpp]: Only difference wrt `myLibIconLaplace.cpp` is here right?

File: dawn/prototype/mylib.hpp [hunk 4]
```diff
@@ -130,7 +136,19 @@ class Grid {
     for(int j = 0; j < (periodic ? ny : ny + 1); ++j)
       for(int i = 0; i < (periodic ? nx : nx + 1); ++i) {
         auto& v = vertex_at(i, j);
-        v = Vertex(i, j, &v - vertices_.data());
+        double px = i;
+        double py = j;
+        if(lx != 0.) {
+          px = double(i) / (nx)*lx;
```
[the remark this answers is not in the document]
Comment (mroethlin) [reply dawn/prototype/mylib.hpp]: yes, properly documented now

File: dawn/prototype/mylib.hpp [hunk 5]
```diff
@@ -130,7 +136,19 @@ class Grid {
...
+        }
+        if(ly != 0.) {
+          py = double(j) / (ny)*ly;
+        }
+        if(equilat) {
+          px = px - 0.5 * py;
+          py = py * sqrt(3) / 2.;
+        }
```
[the remark this answers is not in the document]
Comment (mroethlin) [reply dawn/prototype/mylib.hpp]: issues a warning now

File: dawn/test/unit-test/dawn-c/TestCompiler.cpp [hunk 6]
```diff
@@ -338,4 +338,75 @@ TEST(CompilerTest, DISABLED_NestedReduceField) {
...
+                                      b.reduceOverNeighborExpr(Op::plus, b.at(div_vec), b.lit(0.),
+                                                               dawn::ast::LocationType::Edges,
+                                                               dawn::ast::LocationType::Cells,
+                                                               std::vector<double>{-1, 1}))),
+                  b.stmt(b.assignExpr(
+                      b.at(nabla2t2_vec),
+                      b.binaryExpr(b.at(nabla2t2_vec), b.at(dual_edge_length), Op::divide))),
+                  b.stmt(b.assignExpr(
+                      b.at(nabla2_vec),
+                      b.binaryExpr(b.at(nabla2t1_vec), b.at(nabla2t2_vec), Op::minus))))))));
```
[the remark this answers is not in the document]
Comment (mroethlin) [reply dawn/test/unit-test/dawn-c/TestCompiler.cpp]: correct, seems like our geometric factor is inverted in this case when compared to the ICON grid

File: dawn/prototype/mylibIconLaplace.cpp [hunk 7]
```diff
@@ -0,0 +1,597 @@
...
+
+  auto wave = [](double px, double py) { return sin(px) * sin(py); };
+  auto constant = [](double px, double py) { return 1.; };
+  auto lin = [](double px, double py) { return px; };
+
+  // init zero and test function
+  for(const auto& e : mesh.edges()) {
+    auto [xm, ym] = EdgeMidpoint(e);
+    double py = 2 / sqrt(3) * ym;
+    double px = xm + 0.5 * py;
```
[the remark this answers is not in the document]
Comment (mroethlin) [reply dawn/prototype/mylibIconLaplace.cpp]: this trafo shifts between right and equilateral triangles

File: dawn/prototype/mylibIconLaplace.cpp (hunk 1 above)
Comment (mroethlin) [reply dawn/prototype/mylibIconLaplace.cpp]: thanks 🤦‍♂️

File: dawn/prototype/mylibIconLaplace.cpp (hunk 2 above)
Comment (mroethlin) [reply dawn/prototype/mylibIconLaplace.cpp]:
```
    ! Vorticity should have the right sign
    SELECT CASE (ptr_patch%geometry_info%cell_type)
    CASE (3)
      ifac = 1
    CASE (6)
      ifac = -1
    END SELECT
```
-1 for hexagons, +1 for triangles, we are save ;)

File: dawn/prototype/mylibIconLaplaceDriver.cpp (hunk 3 above)
Comment (mroethlin) [reply dawn/prototype/mylibIconLaplaceDriver.cpp]: right

**Pull request merged**